packages/react-reconciler/src/__tests__/ReactHooksWithNoopRenderer-test.js

```diff
@@ -2682,6 +2684,355 @@ describe('ReactHooksWithNoopRenderer', () => {
     });
   });
 
+  describe('useInsertionEffect', () => {
+    // @gate experimental || www
+    it('fires insertion effects before layout effects', () => {
+      let committedText = '(empty)';
+
+      function Counter(props) {
+        useInsertionEffect(() => {
+          Scheduler.unstable_yieldValue(
+            `Create insertion [current: ${committedText}]`,
+          );
+          committedText = props.count + '';
+          return () => {
+            Scheduler.unstable_yieldValue(
+              `Destroy insertion [current: ${committedText}]`,
+            );
+          };
+        });
+        useLayoutEffect(() => {
+          Scheduler.unstable_yieldValue(
+            `Create layout [current: ${committedText}]`,
+          );
+          return () => {
+            Scheduler.unstable_yieldValue(
+              `Destroy layout [current: ${committedText}]`,
+            );
+          };
+        });
+        useEffect(() => {
+          Scheduler.unstable_yieldValue(
+            `Create passive [current: ${committedText}]`,
+          );
+          return () => {
+            Scheduler.unstable_yieldValue(
+              `Destroy passive [current: ${committedText}]`,
+            );
+          };
+        });
+        return null;
+      }
+      act(() => {
+        ReactNoop.render(<Counter count={0} />);
+
+        expect(Scheduler).toFlushUntilNextPaint([
+          'Create insertion [current: (empty)]',
+          'Create layout [current: 0]',
+        ]);
+        expect(committedText).toEqual('0');
+      });
+
+      expect(Scheduler).toHaveYielded(['Create passive [current: 0]']);
+
+      // Unmount everything
+      act(() => {
+        ReactNoop.render(null);
+
+        expect(Scheduler).toFlushUntilNextPaint([
+          'Destroy insertion [current: 0]',
+          'Destroy layout [current: 0]',
+        ]);
+      });
+
+      expect(Scheduler).toHaveYielded(['Destroy passive [current: 0]']);
+    });
```

Comment:
The one test case I didn't see but we should probably have is where `getSnapshotBeforeUpdate` can read the old style before `useInsertionEffect` mutates it

Reply:
Test added 👍